src/ws/session-utils.c

```diff
@@ -567,3 +767,67 @@ authorize_logger (const char *data)
 {
   warnx ("%s", data);
 }
+
+FILE *
+open_memfd (const char *name)
+{
+  int fd = memfd_create ("cockpit login messages", MFD_ALLOW_SEALING);
+
+  if (fd == -1)
+    return NULL;
+
+  return fdopen (fd, "w");
+}
+
+bool
+seal_memfd (FILE *memfd)
+{
+  if (fflush (memfd) != 0)
+    return false;
+
+  const int seals = F_SEAL_SHRINK | F_SEAL_GROW | F_SEAL_WRITE;
+  return fcntl (fileno (memfd), F_ADD_SEALS, seals) == 0;
+}
+
+/* signal- and after-fork()-safe function to remap file descriptors
+ * according to a specified array.  All other file descriptors are
+ * closed.
```

Comment:
"All other file descriptors are closed" is wrong, no?  They stay open, afaics.

Reply:
Right, I forgot to adjust the comment, or we move the fd closing back into that function (not fussed either way).

Reply:
Ah, now there is closefrom_via_proc.